src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.Cache.cs

```diff
@@ -582,18 +587,34 @@ internal void InitializePropCache()
             }
 
             JsonPropertyInfo[] array = PropInitFunc(context);
-            var properties = new JsonPropertyDictionary<JsonPropertyInfo>(Options.PropertyNameCaseInsensitive, array.Length);
+            Dictionary<string, JsonPropertyInfo>? ignoredMembers = null;
+            JsonPropertyDictionary<JsonPropertyInfo> propertyCache = new(Options.PropertyNameCaseInsensitive, array.Length);
+
             for (int i = 0; i < array.Length; i++)
             {
-                JsonPropertyInfo property = array[i];
-                if (!properties.TryAdd(property.NameAsString, property))
+                JsonPropertyInfo jsonPropertyInfo = array[i];
+                bool hasJsonInclude = jsonPropertyInfo.SrcGen_HasJsonInclude;
+
+                if (!jsonPropertyInfo.SrcGen_IsPublic)
+                {
+                    if (hasJsonInclude)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_JsonIncludeOnNonPublicInvalid(jsonPropertyInfo.ClrName!, jsonPropertyInfo.DeclaringType);
```

Comment:
Is this also an error for fast path?

Reply:
Yup, I've added the logic & tests to account for this.